File: IREngineTestBed/IRGame/scripts/game/character.rb
#!ruby19
# encoding: utf-8
require "helpers"
require "logging"
require "IRGame.dll"

include IRGame::Common
include IRGame::Common::Enums
include IRGame::Characters

class Character
    def Character.build_abc(a_coeffs, b_coeffs, ts)
        a1, a2, a3 = a_coeffs
        b1, b2 = b_coeffs
        
        t1 = 2*a1
        t2 = a3*ts**2
        t3 = a2*ts**2
        
        a = Vector4.new
        a.v1, a.v2, a.v3, a.v4 = [-(t2-t1)/t1,
                                   (t1*ts-t3)/t1,
                                  -(a3*t1*ts-a2*t2)/(a1*t1),
                                  -(a2*t1*ts-a2*t3+a1*t2-a1*t1)/(a1*t1)]
        
        b = Vector2.new
        b.x, b.y = [ ts**2/2,
                     (t1*ts-t3)/t1]
        
        c = Vector2.new
        c.x, c.y = [b1, b2] #[b1/ts, b2/ts]
        
        [a, b, c]
    end
    def Character.build_discrete_model(ax, ay, bx, by, ts)
        # TODO: add parameters check
                
		puts "\n***   Starting build of discrete model...\n" if $verbose
        character = Character.new
        
        character.state[StateVar.XPosStiff] = 1.0
        character.state[StateVar.YPosStiff] = 1.0
        character.state[StateVar.TimeSample] = ts
        
        x_trans, x_ctrl, x_meas = Character.build_abc(ax, bx, ts)        
        y_trans, y_ctrl, y_meas = Character.build_abc(ay, by, ts)
        IRE.log "Created ABC:\ncx: #{x_meas.inspect}\ncy: #{y_meas.inspect}\n" if $verbose
        
        character.set_transition_matrix(x_trans, y_trans)
        character.set_control_matrix(x_ctrl, y_ctrl)
        character.set_measure_matrix(x_meas, y_meas)
        
        character.set_position 0, 0
        character.set_velocity 0, 0
        character.set_orientation 0
        character.set_target_position 0, 0
        character.apply_force 0, 0
        character.state[StateVar.TimeRemainder] = 0
        character.state[StateVar.XForceThresh] = 0
        character.state[StateVar.YForceThresh] = 5
        character.state[StateVar.LastUx] = 0.0
        character.state[StateVar.LastUy] = 0.0
        character.state[StateVar.XPhaseVector] = Vector2.new
        character.state[StateVar.YPhaseVector] = Vector2.new
        
        IRE.log "Character created successfully!\nInitial state:\n#{character.state_to_string}\n" if $verbose
        
        character
    end

    def Character.update(state, time_delta)
        ts = state[StateVar.TimeSample]
        rem = state[StateVar.TimeRemainder]
        state[StateVar.TimeRemainder] = t = rem + time_delta
        if t < ts
            return
        end
                
        n, rem = t.divmod ts
        state[StateVar.TimeRemainder] = rem
        
        n.times{|i|
            Character.step(state)
        }
    end

    def Character.step(state)
        IRE.log "Step started" if $verbose
        
        ts = state[StateVar.TimeSample]
        x_stiff, y_stiff = state[StateVar.XPosStiff], state[StateVar.YPosStiff]
        x, y = state[StateVar.XPhaseVector], state[StateVar.YPhaseVector]
        IRE.log "Recieved phase vectors:\nx: #{x.inspect}\ny: #{y.inspect}\n" if $verbose
        
        px, py, tx, ty = state[StateVar.XPosition], state[StateVar.YPosition], state[StateVar.XPosTarget], state[StateVar.YPosTarget]
        IRE.log "Position vars recieved" if $verbose
        
        fx, fy, fxt, fyt =  state[StateVar.XForce],
                            state[StateVar.YForce],
                            state[StateVar.XForceThresh], 
                            state[StateVar.YForceThresh]
        IRE.log "Force vars recieved" if $verbose
        
        ax, bx, cx = state[StateVar.XTransitionMtx], state[StateVar.XControlMtx], state[StateVar.XMeasureMtx]
        ay, by, cy = state[StateVar.YTransitionMtx], state[StateVar.YControlMtx], state[StateVar.YMeasureMtx]
        
        IRE.log "Model vars recieved" if $verbose
        
        a1, a2, a3, a4 = ax.v1, ax.v2, ax.v3, ax.v4        
        x0 = a1*x.x + a2*x.y
        x1 = a3*x.x + a4*x.y
        
        fx = 0.0 if fx < fxt
        
        ux = fx + x_stiff * (tx - px)
        state[StateVar.LastUx] = ux
        
        x0 += bx.x * ux
        x1 += bx.y * ux
        new_phase = Vector2.new
        new_phase.x, new_phase.y = x0, x1
        
        state[StateVar.XPhaseVector] = new_phase
        
        dpx = cx.x * x0 + cx.y * x1
        pos = state[StateVar.XPosition] = state[StateVar.XPosition] + dpx
        vx = state[StateVar.XVelocity] = dpx/ts
        
        a1, a2, a3, a4 = ay.v1, ay.v2, ay.v3, ay.v4        
        y0 = a1*y.x + a2*y.y
        y1 = a3*y.x + a4*y.y
        
        fy = 0.0 if fy < fyt
        
        uy = fy + y_stiff * (ty - py)
        state[StateVar.LastUy] = uy
        
        y0 += by.x * uy
        y1 += by.y * uy
        new_phase = Vector2.new
        new_phase.x, new_phase.y = y0, y1
        
        state[StateVar.YPhaseVector] = new_phase
        
        dpy = cy.x * y0 + cy.y * y1
        pos = state[StateVar.YPosition] = state[StateVar.YPosition] + dpy
        vy = state[StateVar.YVelocity] = dpy/ts
        
        vel = Vector2.new
        vel.x, vel.y = vx, vy
        state[StateVar.Orientation] = vel.angle.to_deg
        IRE.log "Step end!" if $verbose
    end
    
    def print
        x, y, tx, ty = state[StateVar.XPosition], state[StateVar.YPosition], state[StateVar.XPosTarget], state[StateVar.YPosTarget]
        vx, vy = state[StateVar.XVelocity], state[StateVar.YVelocity]
        
        IRE.log "Position: {#{x}; #{y}}. Target: {#{tx};#{ty}}."
        IRE.log "Velocity: {#{vx}; #{vy}}."
    end
end

File: IREngineTestBed/IRGame/Characters/Character.cs
using IRGame.Common;
using IRGame.Common.Enums;
using System.Text;

namespace IRGame.Characters
{
    public class Character : GameObject
    {
        #region Fields
        private object _statePrintLock = new object();
        #endregion
        #region Ctors

        public Character()
        {
            ApplyForce(0.0, 0.0);
            ApplyMoment(0.0, 0.0);
        }

        #endregion

        #region Public Methods

        public void SetPosition(double x, double y)
        {
            State[StateVar.XPosition] = x;
            State[StateVar.YPosition] = y;
        }
        public void SetVelocity(double x, double y)
        {
            State[StateVar.XVelocity] = x;
            State[StateVar.YVelocity] = y;
        }

        public void SetPosition(Vector2 position)
        {
            SetPosition(position.X, position.Y);
        }
        public void SetVelocity(Vector2 velocity)
        {
            SetVelocity(velocity.X, velocity.Y);
        }
        public void SetOrientation(double angle)
        {
            State[StateVar.Orientation] = angle;
        }

        public Vector2 GetPosition()
        {
            return new Vector2 { X = (double)State[StateVar.XPosition], Y = (double)State[StateVar.YPosition] };
        }
        public Vector2 GetVelocity()
        {
            return new Vector2 { X = (double)State[StateVar.XVelocity], Y = (double)State[StateVar.YVelocity] };
        }
        public double GetOrientation()
        {
            return (double)State[StateVar.Orientation];
        }

        public void ApplyForce(Vector2 force)
        {
            ApplyForce(force.X, force.Y);
        }
        public void ApplyForce(double fx, double fy)
        {
            double fx0 = (double)State.GetOrAdd(StateVar.XForce, 0.0);
            double fy0 = (double)State.GetOrAdd(StateVar.YForce, 0.0);
            
            State.AddOrUpdate(StateVar.XForce, fx, (k, v) => fx);
            State.AddOrUpdate(StateVar.XForceGrad, fx - fx0, (k, v) => fx - fx0);
            State.AddOrUpdate(StateVar.XForce, fy, (k, v) => fy);
            State.AddOrUpdate(StateVar.YForceGrad, fy - fy0, (k, v) => fy - fy0);
        }
        
        public void ApplyMoment(Vector2 moment)
        {
            ApplyMoment(moment.X, moment.Y);
        }
        public void ApplyMoment(double mx, double my)
        {
            State[StateVar.XMoment] = mx;
            State[StateVar.YMoment] = my;
        }

        public void SetTargetPosition(Vector2 target)
        {
            SetTargetPosition(target.X, target.Y);
        }
        public void SetTargetPosition(double xTarget, double yTarget)
        {
            State[StateVar.XPosTarget] = xTarget;
            State[StateVar.YPosTarget] = yTarget;
        }
        
        public void SetTargetVelocity(Vector2 target)
        {
            SetTargetVelocity(target.X, target.Y);
        }
        public void SetTargetVelocity(double xTarget, double yTarget)
        {
            State[StateVar.XVelTarget] = xTarget;
            State[StateVar.YVelTarget] = yTarget;
        }

        public void SetTransitionMatrix(Vector4 xMtxElements, Vector4 yMtxElements)
        {
            State[StateVar.XTransitionMtx] = xMtxElements;
            State[StateVar.YTransitionMtx] = yMtxElements;
        }

        public void SetControlMatrix(Vector8 xMtxElements, Vector8 yMtxElements)
        {
            State[StateVar.XControlMtx] = xMtxElements;
            State[StateVar.YControlMtx] = yMtxElements;
        }

        public string StateToString()
        {
            // TODO: wrap all operations with State dictionary to lock
            var builder = new StringBuilder();
            lock (_statePrintLock)
            {
                foreach (var key in State.Keys)
                {
                    builder.AppendFormat("{0} => {1}\n", key, State[key]);
                }
            }
            return builder.ToString();
        }
        #endregion
    }
}
